packages/eslint-plugin-ckeditor5-rules/lib/rules/ckeditor-plugin-flags.js

```diff
@@ -0,0 +1,680 @@
+/**
```

Comment:
Does this rule really need almost 700 lines?

IMO, we don't need code that ensures that these two getters are defined after the plugin name or in proper order. We just need to know if they are there (when required) and if their value is right.

A similar rule has a little over 100 lines and is much easier to reason about: https://github.com/ckeditor/ckeditor5-linters-config/blob/master/packages/eslint-plugin-ckeditor5-rules/lib/rules/require-as-const-returns-in-methods.js. It doesn't have a fixer, but new plugins are added so rarely that we can live without one.

Reply:
Yes, it needs it, as far as we want to have auto fixing which drastically improves ergonomics of the code.

Reply:
I'd still argue that we don't need a fixer or anything fancy, like order detection.

Again, new plugins are introduced very rarely and when they are, copying these two getters from an existing plugin is a matter of 15 seconds — not much of a DX win to have it auto-fix. Actually, running `yarn lint --fix` would probably take longer, considering how large is our codebase.

On the other hand, having a complex eslint plugin has a higher maintenance cost. Not to mention a performance cost, as this rule will be executed on every plugin, on every CI run, and anytime we try to push our changes to GitHub (because of the `lint-staged`).

IMO, what we need is simple detection if these two getters are defined when they are supposed to and if the returned value is correct. Nothing more, nothing less.

Reply:
@filipsobol 

> I'd still argue that we don't need a fixer or anything fancy, like order detection.

You are right, these things do not need fixer. Basically, we can say exactly the same about every ESLint rule we have, and every other ESLint defined somewhere else in the world. The problem is that some of these rules improves DX of codebase for a tiny fraction of the time, making contributions a little bit faster, and use less brain power to think about placement of these methods and proper typing.   

The real problem is that these methods **must** be defined in correct format, in order to be properly detected by some of our data usage helpers to keep compatibility with our building flow. They should be `getters`, must be `public` and it'd be also the best option to return proper typing, not just `boolean` in case of assertions. Unfortunately, TS does not provide any solution for enforcing such things in the typing system, simple detection of presence of these methods is insufficient and is error-prone.

I'm also a huge fan of being consistent across our codebase and being super restrictive about that, just to reduce comments in PRs.

> Again, new plugins are introduced very rarely and when they are, copying these two getters from an existing plugin is a matter of 15 seconds — not much of a DX win to have it auto-fix. Actually, running yarn lint --fix would probably take longer, considering how large is our codebase.

It's true, while using `yarn lint --fix` method, which is the slowest one to fix this kind of issues. Most IDEs today have a built-in auto-fixer and the whole feature was designed to be as fast as it should be to use. In other words, ppm -> fix, solves problem in less than one second. It's not 15 seconds, as you said. 

> On the other hand, having a complex eslint plugin has a higher maintenance cost.

No, it's not. ESlint updates does not happen every day, and I'm pretty sure, that this one particular change will stay here for _months_ or even _years_, without any further modification. I kept in mind that some maintenance will be needed anyway - during ESLint upgrade, so I left comments explaining what's already happening in the code, and the code has been written in mind to be as most expressive as it can be.

> Not to mention a performance cost, as this rule will be executed on every plugin, on every CI run

I checked that, and it turns out it's not a problem. We have something like 400 plugins in our public source tree. The code of the rule is designed to be highly JIT-able, as there are no currying and other fancy FP techniques.

![obraz](https://github.com/user-attachments/assets/86d7fc47-93fe-413e-b7f1-1d00582996c8)

The performance cost is approximately 500ms, in one run. I'd have to create more proper testing, but I'd not expect any visible impact on CI. 

> and anytime we try to push our changes to GitHub (because of the lint-staged).

`lint-staged` should run on staged file, with only few added plugins / files. So it'd say it's extra 50-100ms on average.

> IMO, what we need is simple detection if these two getters are defined when they are supposed to and if the returned value is correct. Nothing more, nothing less.

We have to add proper `eslint` rules for disallowing certain flags from plugins (and proper autofixing it) for our development tools and packages exported from package generator.  

---

It's most likely that this discussion exceeded the whole lifetime support time for this rule. I'd suggest staying as it's and observe if there are any problems with current solution. If there are - let's fix them, and if there are too many of issues - simplify the rule.